File: v2/java/random/Boj1967.java
package random;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Boj1967 {
    static BufferedReader br;
    static StringTokenizer st;
    static int N, maxDist, maxNode;
    static boolean[] visited;
    static ArrayList<ArrayList<int[]>> grp;

    public static void main(String[] args) throws IOException {
        br = new BufferedReader(new InputStreamReader(System.in));
        N = Integer.parseInt(br.readLine());

        visited = new boolean[N + 1];
        grp = new ArrayList<>();

        for (int i = 0; i <= N; i++) {
            grp.add(new ArrayList<>());
        }

        for (int i = 1; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            int c = Integer.parseInt(st.nextToken());

            grp.get(a).add(new int[] { b, c });
            grp.get(b).add(new int[] { a, c });
        }

        dfs(1, 0);
        Arrays.fill(visited, false);
        dfs(maxNode, 0);

        System.out.println(maxDist);
    }

    static void dfs(int node, int sum) {
        visited[node] = true;

        if (sum > maxDist) {
            maxNode = node;
            maxDist = sum;
        }

        for (int[] tmp : grp.get(node)) {
            if (!visited[tmp[0]]) {
                visited[tmp[0]] = true;
                dfs(tmp[0], sum + tmp[1]);
            }
        }
    }
}

File: v1/review/kotlin/src/main/kotlin/soma/boj_1967.kt
package soma

private lateinit var tree: List<ArrayList<Pair<Int, Int>>>
private lateinit var visited: BooleanArray
private var maxDist = 0
private var farNode = 0

private fun dfs(start: Int, sum: Int) {
    visited[start] = true

    if (sum > maxDist) {
        farNode = start
        maxDist = sum
    }

    for ((node, weight) in tree[start]) {
        if (!visited[node]) {
            visited[node] = true
            dfs(node, sum + weight)
        }
    }
}

fun boj_1967() = with(System.`in`.bufferedReader()) {
    val n = readLine().toInt()
    tree = List(n + 1) { arrayListOf() }
    visited = BooleanArray(n + 1) { false }

    for (i in 1 until n) {
        val (p, c, w) = readLine().split(' ').map { it.toInt() }
        tree[p].add(Pair(c, w))
        tree[c].add(Pair(p, w))
    }

    dfs(1, 0)
    visited.fill(false)
    dfs(farNode, 0)

    println(maxDist)
}